### PR TITLE
arch: arc: support arc secureshield to build/test/debug with west commands

### PR DESCRIPTION
**Pull request by YuguoWH**
ARC secureshield is a set of features supporting execution of applications in isolated, trusted execution environments. Previously in the scope of secureshield, the sample at `zephyr/samples/boards/arc_secure_services` can be used as a secure image and runs with a non-secure zephyr. However there's no convenient way to build/test/debug it but instead use several make commands. This PR aims to improve that. 
It is supported to use west command to build/test/debug just like other boards. The PR also simply the sample to non-multithreading mode for less overhead. 
Currently the changes are applied to nsim_sem target only, there will be support to other targets with secureshield (e.g. emsdp, emsk) soon

### CODE REVIEW

File: arch/arc/CMakeLists.txt [hunk 1]
```diff
@@ -20,3 +20,28 @@ if(CONFIG_ISA_ARCV2)
...
+
+
+if (CONFIG_BUILD_WITH_SECURESHIELD)
+  set(SECURESHIELD_BINARY_DIR ${CMAKE_BINARY_DIR}/secureshield)
+  file(MAKE_DIRECTORY ${SECURESHIELD_BINARY_DIR})
+  add_custom_target(secureshield_cmake
+    COMMAND west build -p auto -b ${CONFIG_SECURESHIELD_BOARD} -d ${SECURESHIELD_BINARY_DIR} ${ZEPHYR_BASE}/${CONFIG_SECURESHIELD_IMAGE_PATH}
+    WORKING_DIRECTORY ${SECURESHIELD_BINARY_DIR}
+    COMMAND_EXPAND_LISTS
+  )
```
Comment (mbolivar-nordic) [inline arch/arc/CMakeLists.txt:33]: I'm really not sure about this. I think @tejlmand should confirm this is OK. I know we do some things with sub-images in TF-M (so cc @ioannisg and @microbuilder for help here) but I'm not sure if this style of calling out to west build from the build system itself is something we should encourage.

@nashif -- any feedback?

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Putting a -1 on this until we can discuss the sub-image question.

File: arch/arc/CMakeLists.txt (hunk 1 above)
Comment (ruuddw) [reply arch/arc/CMakeLists.txt:33]: Marti, can you explain your concern? The goal here is to make it easy for users to combine building/flashing/debugging a secure and a normal binary. In the TF-M case, the secure binary is build as an external module I believe. For ARC we decided to have a more closer Zephyr integration and use west to build the secure binary.
Do you have a suggestion for a different (better?) way of achieving the same (we want to e.g. be able to leverage twister and CI etc. instead of current manual steps for combining the secure and normal images)?

**Review by ruuddw — APPROVED** [review]
Looks fine to me, but let's have some broader review to align with other "Secure mode" (TF-M) approach, as Marti suggests.

File: arch/arc/CMakeLists.txt (hunk 1 above)
Comment (mbolivar-nordic) [reply arch/arc/CMakeLists.txt:33]: @ruuddw the topic of building multiple images within the zephyr build system has a long and contentious history. I won't go into all the details in this comment -- there was even an entire BoF talk about it at ELCE a few years ago to discuss a proposal from Nordic on it which was ultimately rejected by the Zephyr community: https://www.youtube.com/watch?v=ZZJ-r94MUnI

Given that history I think it's worth making sure everyone knows about and is in agreement with the approach here.

**Comment by abrodkin** [comment]
@mbolivar @nashif maybe it worth discussion on of the calls (though I'm not sure which one that should be)?

File: arch/arc/CMakeLists.txt (hunk 1 above)
Comment (ruuddw) [reply arch/arc/CMakeLists.txt:33]: @mbolivar, thanks for the pointer/reminder. Unfortunately the discussion is hard to follow in the recording.
The approach here isn't following the 'nonrecursive cmake' which was challenged to introduce too much complexity to handle the namespaces. It follows the 'recursive cmake' - we'll pay extra attention to handling dependencies, but I think we'll be fine with only limited dependencies in this case.

**Comment by mbolivar-nordic** [comment]
> maybe it worth discussion on of the calls (though I'm not sure which one that should be)?

@abrodkin great idea; let's add this to the dev-review agenda for this week.

**Comment by galak** [comment]
dev-review (Jul 29, 2021):
* Don't use west to do the second build, this should be cmake if anything
* We need a more generalized solution for multi-image building that is not using the zephyr build system.  Nordic has something for their downstream SDK that might be worth looking at (@mbolivar-nordic, @carlescufi).  Also possibly a west extension or using cmake for this purpose.

**Comment by mbolivar-nordic** [comment]
>     * Nordic has something for their downstream SDK that might be worth looking at (@mbolivar-nordic, @carlescufi)

The core of this is here:

https://github.com/nrfconnect/sdk-nrf/blob/c239a38d97463e7fe8c755b1c81c7eaf9753f027/cmake/multi_image.cmake#L42

**Comment by YuguoWH** [comment]
Now use cmake instead of west for second image build

**Comment by mbolivar-nordic** [comment]
> Now use cmake instead of west for second image build

Thank you for your update. However, I think the consensus at dev-review was that this should not be merged even with that change made as we do not want even recursive builds of different west binaries happening from the same build system.

**Comment by ruuddw** [comment]
> > Now use cmake instead of west for second image build
> 
> Thank you for your update. However, I think the consensus at dev-review was that this should not be merged even with that change made as we do not want even recursive builds of different zephyr binaries happening from the same build system, with west or not.

Indeed, but a generic solution for multi-image handling will take some time. This at least addresses the major concern that there should not be a dependency from cmake to west, only vice versa. Let's take it step by step.

**Comment by mbolivar-nordic** [comment]
> This at least addresses the major concern that there should not be a dependency from cmake to west, only vice versa. Let's take it step by step.

Sure, as long as "take it step by step" does not imply this gets merged as-is. I am sorry to be stubborn about this, but I think that was the clear outcome from the dev-review discussion.

**Comment by ruuddw** [comment]
@mbolivar-nordic I added the DNM label to be explicit on this. Dev-review feedback that Kumar listed:

1. Don't use west to do the second build, this should be cmake if anything
2. We need a more generalized solution for multi-image building that is not using the zephyr build system. Nordic has something for their downstream SDK that might be worth looking at (@mbolivar-nordic, @carlescufi). Also possibly a west extension or using cmake for this purpose.

1 has been addressed, 2 still needs to be done. For this specific "ARC SecureShield" case (multiple images, but conceptually it is one, single processor running both images as if they are one), we'll consider the best options and come with a new proposal.

**Review by tejlmand — CHANGES_REQUESTED** [review]
a couple of small observations.

I think i'm aligned with @mbolivar-nordic and the outcome from the dev-review meeting.

One more thing to consider if doing a recursive Zephyr build / multi-image which is done in Nordic SDK is the configuration of the sub-image.

When external projects are used elsewhere in Zephyr, the needed settings from the external project has corresponding mappings into the Zephyr Kconfig system, for example TF-M:
https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/trusted-firmware-m/Kconfig.tfm
https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/trusted-firmware-m/CMakeLists.txt#L33-L102

This is not covered in this PR.
